File: yacoo_rpg/src/game/combat.ts
import { ATTACK_CATEGORIES, HANDS } from './constants';
import { getHeroStats } from './equipment';
import type { DieValue, EnemyStats, EquipmentSet, RunState, YahtzeeAttackCategory, YahtzeeHand } from './types';

export function createEnemy(stage: number): EnemyStats {
  // Rebalanced for turn-based combat (M1.5).
  // Target: ~5–8 turns per fight at default equipment.
  // attack scaled so hero takes ~25–40% max HP per turn without upgrades.
  return {
    stage,
    maxHp: 100 + 40 * (stage - 1),
    attack: 18 + 4 * (stage - 1),
    name: stage % 5 === 0 ? 'Moss Boss' : `Forest Grump ${stage}`
  };
}

export function clampHp(value: number): number {
  return Math.max(0, Math.floor(value));
}

export function calculateDiceDamage(equipment: EquipmentSet, hand: YahtzeeHand): number {
  const hero = getHeroStats(equipment);
  return Math.floor(hero.attack * HANDS[hand].multiplier * (1 + hero.diceBonus));
}

export function calculateSelectedCategoryDamage(equipment: EquipmentSet, category: YahtzeeAttackCategory): number {
  const hero = getHeroStats(equipment);
  const metadata = ATTACK_CATEGORIES.find((item) => item.category === category);

  if (!metadata) {
    throw new Error(`Unknown attack category: ${category}`);
  }

  return Math.floor(hero.attack * metadata.multiplier * (1 + hero.diceBonus));
}

export function calculateHeroAutoDamage(equipment: EquipmentSet): number {
  const hero = getHeroStats(equipment);
  return hero.attack + hero.autoBonus;
}

export function calculateEnemyDamage(enemy: EnemyStats, equipment: EquipmentSet): number {
  const hero = getHeroStats(equipment);
  return Math.max(1, enemy.attack - hero.defense);
}

/**
 * Ultimate (dice special attack) damage — M1 version.
 * Formula: floor((pipSum + weaponAttack) * handMultiplier * (1 + diceBonus))
 * M3 will add traitDamageBonus, specialtyMultiplier, treasure effects.
 */
export function calculateUltimateDamage(args: {
  dice: DieValue[];
  hand: YahtzeeHand;
  equipment: EquipmentSet;
  run?: RunState;
}): number {
  const { dice, hand, equipment } = args;
  const hero = getHeroStats(equipment);
  const pipSum = dice.reduce((sum, d) => sum + d, 0);
  const handMeta = HANDS[hand];
  return Math.floor((pipSum + hero.attack) * handMeta.multiplier * (1 + hero.diceBonus));
}

export function calculateUltimateCategoryDamage(args: {
  dice: DieValue[];
  category: YahtzeeAttackCategory;
  equipment: EquipmentSet;
  run?: RunState;
}): number {
  const { dice, category, equipment } = args;
  const metadata = ATTACK_CATEGORIES.find((item) => item.category === category);
  if (!metadata) throw new Error(`Unknown attack category: ${category}`);
  const hero = getHeroStats(equipment);
  const pipSum = dice.reduce((sum, d) => sum + d, 0);
  return Math.floor((pipSum + hero.attack) * metadata.multiplier * (1 + hero.diceBonus));
}

File: yacoo_rpg-android/app/src/main/java/com/yacoo/rpg/game/Combat.kt
package com.yacoo.rpg.game

import kotlin.math.floor
import kotlin.math.max

fun createEnemy(stage: Int): EnemyStats = EnemyStats(
    stage  = stage,
    maxHp  = 100 + 40 * (stage - 1),
    attack = 18  + 4  * (stage - 1),
    name   = if (stage % 5 == 0) "Moss Boss" else "Forest Grump $stage"
)

fun clampHp(value: Int): Int = max(0, value)

fun calculateDiceDamage(equipment: EquipmentSet, hand: YahtzeeHand): Int {
    val hero = getHeroStats(equipment)
    return floor(hero.attack * hand.multiplier * (1 + hero.diceBonus)).toInt()
}

fun calculateSelectedCategoryDamage(equipment: EquipmentSet, category: YahtzeeAttackCategory): Int {
    val hero = getHeroStats(equipment)
    return floor(hero.attack * category.multiplier * (1 + hero.diceBonus)).toInt()
}

fun calculateHeroAutoDamage(equipment: EquipmentSet): Int {
    val hero = getHeroStats(equipment)
    return hero.attack + hero.autoBonus
}

fun calculateEnemyDamage(enemy: EnemyStats, equipment: EquipmentSet): Int {
    val hero = getHeroStats(equipment)
    return max(1, enemy.attack - hero.defense)
}

fun calculateUltimateDamage(
    dice: List<DieValue>,
    hand: YahtzeeHand,
    equipment: EquipmentSet
): Int {
    val hero    = getHeroStats(equipment)
    val pipSum  = dice.sum()
    return floor((pipSum + hero.attack) * hand.multiplier * (1 + hero.diceBonus)).toInt()
}

fun calculateUltimateCategoryDamage(
    dice: List<DieValue>,
    category: YahtzeeAttackCategory,
    equipment: EquipmentSet
): Int {
    val hero   = getHeroStats(equipment)
    val pipSum = dice.sum()
    return floor((pipSum + hero.attack) * category.multiplier * (1 + hero.diceBonus)).toInt()
}
